File: examples/runnable/11-determinism-replay/go/pipeline.go
// Package main holds the determinism-replay lab: a data pipeline Workflow plus a
// "harmless looking" reordered refactor. The Go equivalent of the Java
// determinism-replay lab. The original extracts THEN loads; the reordered version
// loads THEN extracts. Both register under the SAME Workflow type name, so a
// history recorded from the original replays against the reordered code — and the
// different command order trips a non-determinism error. That regression is
// exactly what replay testing catches.
package main

import (
	"context"
	"time"

	"go.temporal.io/sdk/workflow"
)

const (
	TaskQueue    = "replay-demo"
	WorkflowName = "DataPipelineWorkflow"
)

// Both impls register under the SAME Workflow type name, so a history recorded
// from one is replayed against the other.
var registerOpts = workflow.RegisterOptions{Name: WorkflowName}

// Extract is an Activity returning the extracted rows.
func Extract(ctx context.Context) (string, error) {
	return "rows:100", nil
}

// Load is an Activity that loads the given data.
func Load(ctx context.Context, data string) (string, error) {
	return "loaded " + data, nil
}

// DataPipelineWorkflow is the original, shipped version: extract THEN load.
func DataPipelineWorkflow(ctx workflow.Context) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
	})
	var extracted string
	if err := workflow.ExecuteActivity(ctx, Extract).Get(ctx, &extracted); err != nil {
		return "", err
	}
	var loaded string
	if err := workflow.ExecuteActivity(ctx, Load, extracted).Get(ctx, &loaded); err != nil {
		return "", err
	}
	return loaded, nil
}

// ReorderedPipelineWorkflow is the breaking refactor: load THEN extract. Same
// type name, new command order.
func ReorderedPipelineWorkflow(ctx workflow.Context) (string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
	})
	var loaded string
	if err := workflow.ExecuteActivity(ctx, Load, "rows:0").Get(ctx, &loaded); err != nil {
		return "", err
	}
	var extracted string
	if err := workflow.ExecuteActivity(ctx, Extract).Get(ctx, &extracted); err != nil {
		return "", err
	}
	return loaded + " / " + extracted, nil
}

File: examples/runnable/11-determinism-replay/python/pipeline.py
"""Data pipeline workflow + a 'harmless looking' reordered refactor.

The Python equivalent of the Java determinism-replay lab. The original Workflow
extracts THEN loads; the reordered version loads THEN extracts. Both share the
same Workflow type name, so a history recorded from the original replays against
the reordered code — and the different command order trips a non-determinism
error. That regression is exactly what replay testing catches.
"""

from datetime import timedelta

from temporalio import activity, workflow

TASK_QUEUE = "replay-demo"
WORKFLOW_NAME = "DataPipelineWorkflow"


@activity.defn
async def extract() -> str:
    return "rows:100"


@activity.defn
async def load(data: str) -> str:
    return f"loaded {data}"


@workflow.defn(name=WORKFLOW_NAME)
class DataPipelineWorkflow:
    """The original, shipped version: extract THEN load."""

    @workflow.run
    async def run(self) -> str:
        opts = dict(start_to_close_timeout=timedelta(seconds=10))
        extracted = await workflow.execute_activity(extract, **opts)
        return await workflow.execute_activity(load, extracted, **opts)


@workflow.defn(name=WORKFLOW_NAME)
class ReorderedPipelineWorkflow:
    """The breaking refactor: load THEN extract — same type name, new command order."""

    @workflow.run
    async def run(self) -> str:
        opts = dict(start_to_close_timeout=timedelta(seconds=10))
        loaded = await workflow.execute_activity(load, "rows:0", **opts)
        extracted = await workflow.execute_activity(extract, **opts)
        return f"{loaded} / {extracted}"
